Lib/json/tool.py

```diff
@@ -22,20 +22,21 @@ def main():
     parser = argparse.ArgumentParser(prog=prog, description=description)
     parser.add_argument('infile', nargs='?', type=argparse.FileType(),
                         help='a JSON file to be validated or pretty-printed')
-    parser.add_argument('outfile', nargs='?', type=argparse.FileType('w'),
+    parser.add_argument('outfile', nargs='?', default='-',
                         help='write the output of infile to outfile')
     parser.add_argument('--sort-keys', action='store_true', default=False,
                         help='sort the output of dictionaries alphabetically by key')
     options = parser.parse_args()
 
     infile = options.infile or sys.stdin
-    outfile = options.outfile or sys.stdout
     sort_keys = options.sort_keys
     with infile:
         try:
             obj = json.load(infile)
         except ValueError as e:
             raise SystemExit(e)
+
+    outfile = sys.stdout if options.outfile == '-' else open(options.outfile, 'w')
```

Comment:
You need to more properly handle IOError:

Example, consider that `/bla` doesn't exist on the system with your version we obtain the following output:
```
$ python tool.py.3 test.json /bla/test.json
Traceback (most recent call last):
  File "tool.py.3", line 46, in <module>
    main()
  File "tool.py.3", line 39, in main
    outfile = sys.stdout if options.outfile == '-' else open(options.outfile, 'w')
IOError: [Errno 2] No such file or directory: '/bla/test.json'
```

With the older version of this tools with obtain the following output:
```
$ python tool.py test.json /bla/test.json
usage: python -m json.tool [-h] [--sort-keys] [infile] [outfile]
python -m json.tool: error: argument outfile: can't open '/bla/test.json': [Errno 2] No such file or directory: '/bla/test.json'
```

Reply:
The following code provide the same output : 
```python
import argparse
import json
import os
import sys


def main():
    prog = 'python -m json.tool'
    description = ('A simple command line interface for json module '
                   'to validate and pretty-print JSON objects.')
    parser = argparse.ArgumentParser(prog=prog, description=description)
    parser.add_argument('infile', nargs='?', type=argparse.FileType(),
                        help='a JSON file to be validated or pretty-printed')
    parser.add_argument('outfile', nargs='?', default='-',
                        help='write the output of infile to outfile')
    parser.add_argument('--sort-keys', action='store_true', default=False,
                        help='sort the output of dictionaries alphabetically by key')
    options = parser.parse_args()

    infile = options.infile or sys.stdin
    try:
        outfile = sys.stdout if options.outfile == '-' else open(options.outfile, 'w')
    except IOError as e:
        parser.print_help(e)
    sort_keys = options.sort_keys
    with infile:
        try:
            obj = json.load(infile)
        except ValueError as e:
            raise SystemExit(e)


    with outfile:
        json.dump(obj, outfile, sort_keys=sort_keys, indent=4)
        outfile.write('\n')


if __name__ == '__main__':
    main()
```

Reply:
```python
try:
    outfile = sys.stdout if options.outfile == '-' else open(options.outfile, 'w')
except IOError as e:
    parser.print_help(e)
```

Reply:
You can implement this or get the fix from my fork:

```
$ git remote add 4383 https://github.com/4383/cpython
$ git fetch 4383
$ git branch fix_json 4383/json.tool-same-infile-outfile
$ git checkout bugfix/json.tool-same-infile-outfile
$ git cherry-pick 1e8b99f 
```

Reply:
Nice catch @4383, the proposed fix does not work because it would opening outfile before reading infile might overwrite its content if both point to the same file.

I did what you proposed after the first context manager, when infile has been closed.

Reply:
@remilapeyre you are welcome :)
My patch on repo was a little bit crap, it was not directly tested, it was a compromise between your solution and a temporary solution... it was more to help you than anything else :)